File: apps/app-frontend/src/helpers/jre.js
/**
 * All theseus API calls return serialized values (both return values and errors);
 * So, for example, addDefaultInstance creates a blank instance object, where the Rust struct is serialized,
 *  and deserialized into a usable JS object.
 */
import { invoke } from '@tauri-apps/api/core'

/*

JavaVersion {
    path: Path
    version: String
}

*/

export async function get_java_versions() {
	return await invoke('plugin:jre|get_java_versions')
}

export async function get_java_default_versions() {
	return await invoke('plugin:jre|get_java_default_versions')
}

export async function set_java_version(javaVersion) {
	return await invoke('plugin:jre|set_java_version', { javaVersion })
}

export async function set_java_default_version(majorVersion, path) {
	return await invoke('plugin:jre|set_java_default_version', { majorVersion, path })
}

export async function remove_java_default_version(majorVersion) {
	return await invoke('plugin:jre|remove_java_default_version', { majorVersion })
}

export async function remove_java_version(path) {
	return await invoke('plugin:jre|remove_java_version', { path })
}

// Finds all the installation of the given Java version, if it exists
// Returns [JavaVersion]
export async function find_filtered_jres(
	version,
	fullScan = false,
	forceFresh = false,
	exhaustive = false,
) {
	return await invoke('plugin:jre|jre_find_filtered_jres', {
		version,
		fullScan,
		forceFresh,
		exhaustive,
	})
}

// Gets java version from a specific path by trying to run 'java -version' on it.
// This also validates it, as it returns null if no valid java version is found at the path
export async function get_jre(path) {
	return await invoke('plugin:jre|jre_get_jre', { path })
}

// Tests JRE version by running 'java -version' on it.
// Returns true if the version is valid, and matches given (after extraction)
export async function test_jre(path, majorVersion) {
	return await invoke('plugin:jre|jre_test_jre', { path, majorVersion })
}

// Automatically installs specified java version
export async function auto_install_java(javaVersion) {
	return await invoke('plugin:jre|jre_auto_install_java', { javaVersion })
}

export async function respond_to_java_download_confirmation(requestId, approved) {
	return await invoke('plugin:jre|jre_respond_to_download_confirmation', {
		requestId,
		approved,
	})
}

export async function list_java_distribution_versions(distribution) {
	return await invoke('plugin:jre|list_java_distribution_versions', { distribution })
}

// Get max memory in KiB
export async function get_max_memory() {
	return await invoke('plugin:jre|jre_get_max_memory')
}

export async function get_memory_status(instanceId, requestedMemoryMb, automatic) {
	return await invoke('plugin:jre|jre_get_memory_status', {
		instanceId,
		requestedMemoryMb,
		automatic,
	})
}

export async function optimize_memory() {
	return await invoke('plugin:jre|jre_optimize_memory')
}

export async function list_java_feed_vendors() {
	return await invoke('plugin:jre|list_java_feed_vendors')
}

export async function list_java_feed_versions(vendor) {
	return await invoke('plugin:jre|list_java_feed_versions', { vendor })
}

export async function download_java_from_feed(vendor, jdkVersionMajor) {
	return await invoke('plugin:jre|download_java_from_feed', { vendor, jdkVersionMajor })
}

export async function download_java(vendor, version) {
	return await invoke('plugin:jre|download_java', { vendor, version })
}

File: apps/app/src/api/jre.rs
use crate::api::Result;
use std::path::PathBuf;
use tauri::plugin::TauriPlugin;
use theseus::prelude::*;

pub fn init<R: tauri::Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::new("jre")
        .invoke_handler(tauri::generate_handler![
            get_java_versions,
            get_java_default_versions,
            set_java_version,
            set_java_default_version,
            remove_java_default_version,
            remove_java_version,
            jre_find_filtered_jres,
            jre_get_jre,
            jre_test_jre,
            jre_auto_install_java,
            jre_respond_to_download_confirmation,
            jre_get_max_memory,
            jre_get_memory_status,
            jre_optimize_memory,
            list_java_distribution_versions,
            list_java_feed_vendors,
            list_java_feed_versions,
            download_java_from_feed,
            download_java,
        ])
        .build()
}

#[tauri::command]
pub async fn get_java_versions() -> Result<Vec<JavaVersion>> {
    Ok(jre::get_java_versions().await?)
}

#[tauri::command]
pub async fn get_java_default_versions() -> Result<Vec<JavaVersion>> {
    Ok(jre::get_java_default_versions().await?)
}

#[tauri::command]
pub async fn set_java_version(java_version: JavaVersion) -> Result<()> {
    jre::set_java_version(java_version).await?;
    Ok(())
}

#[tauri::command]
pub async fn set_java_default_version(
    major_version: u32,
    path: String,
) -> Result<JavaVersion> {
    Ok(jre::set_java_default_version(major_version, path).await?)
}

#[tauri::command]
pub async fn remove_java_default_version(major_version: u32) -> Result<()> {
    jre::remove_java_default_version(major_version).await?;
    Ok(())
}

#[tauri::command]
pub async fn remove_java_version(path: String) -> Result<()> {
    jre::remove_java_version(path).await?;
    Ok(())
}

// Finds the installation of Java 8, if it exists
#[tauri::command]
pub async fn jre_find_filtered_jres(
    version: Option<u32>,
    full_scan: bool,
    force_fresh: bool,
    exhaustive: bool,
) -> Result<Vec<JavaVersion>> {
    Ok(
        jre::find_filtered_jres(version, full_scan, force_fresh, exhaustive)
            .await?,
    )
}

// Validates JRE at a given path
// Returns None if the path is not a valid JRE
#[tauri::command]
pub async fn jre_get_jre(path: PathBuf) -> Result<JavaVersion> {
    Ok(jre::check_jre(path).await?)
}

// Tests JRE of a certain version
#[tauri::command]
pub async fn jre_test_jre(path: PathBuf, major_version: u32) -> Result<bool> {
    Ok(jre::test_jre(path, major_version).await?)
}

// Auto installs java for the given java version
#[tauri::command]
pub async fn jre_auto_install_java(
    java_version: u32,
) -> Result<Option<PathBuf>> {
    Ok(jre::auto_install_java(java_version).await?)
}

#[tauri::command]
pub fn jre_respond_to_download_confirmation(
    request_id: uuid::Uuid,
    approved: bool,
) -> bool {
    jre::respond_to_java_download_confirmation(request_id, approved)
}

#[tauri::command]
pub async fn list_java_distribution_versions(
    distribution: String,
) -> Result<Vec<u32>> {
    Ok(jre::list_java_distribution_versions(distribution).await?)
}

// Gets the maximum memory a system has available.
#[tauri::command]
pub async fn jre_get_max_memory() -> Result<u64> {
    Ok(jre::get_max_memory().await?)
}

#[tauri::command]
pub async fn jre_get_memory_status(
    instance_id: Option<String>,
    requested_memory_mb: u32,
    automatic: bool,
) -> Result<jre::MemoryStatus> {
    Ok(jre::get_memory_status(
        instance_id.as_deref(),
        requested_memory_mb,
        automatic,
    )
    .await?)
}

#[tauri::command]
pub async fn jre_optimize_memory()
-> Result<theseus::memory::MemoryOptimizationResult> {
    Ok(theseus::memory::optimize().await?)
}

#[tauri::command]
pub async fn list_java_feed_vendors() -> Result<Vec<String>> {
    Ok(jre::list_java_feed_vendors().await?)
}

#[tauri::command]
pub async fn list_java_feed_versions(
    vendor: String,
) -> Result<Vec<JdkVersionInfo>> {
    Ok(jre::list_java_feed_versions(&vendor).await?)
}

#[tauri::command]
pub async fn download_java_from_feed(
    vendor: String,
    jdk_version_major: u32,
) -> Result<PathBuf> {
    Ok(jre::download_java_from_feed(&vendor, jdk_version_major).await?)
}

#[tauri::command]
pub async fn download_java(
    vendor: String,
    version: u32,
) -> Result<theseus::install::InstallJobSnapshot> {
    Ok(theseus::install::download_java(vendor, version).await?)
}
